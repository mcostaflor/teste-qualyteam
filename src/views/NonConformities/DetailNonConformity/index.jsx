import React, { useEffect, useState } from 'react';
import Axios from 'axios';
import { Link } from 'react-router-dom';

export default function NonConformity({ history, match }) {

    const [nonConformity, setNonConformity] = useState({});
    const [correctiveActions, setCorrectiveActions] = useState([])


    useEffect(() => {
        Axios.get(`http://localhost:3000/non-conformities/${match.params.id}`)
            .then(res => {
                setNonConformity(res.data);
                setCorrectiveActions(correctiveActions);
            });
    }, [match.params.id]);

    console.log(correctiveActions)

    return (
        <><Link to='/nonconfirmities'>voltar</Link>
            <div>
                <div>
                    {nonConformity.id}
                </div>
                <div>
                    {nonConformity['ocurrence-date']}
                </div>
                <div>
                    {nonConformity.description}
                </div>
            </div>
            <div>
                {correctiveActions.map(item =>
                    <div key={item.id}>
                        <div>{item.id}</div>
                    </div>
                )}
            </div>
        </>
    );
}