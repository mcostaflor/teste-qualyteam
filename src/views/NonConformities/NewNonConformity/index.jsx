import React, { useState, useEffect } from 'react';
import Axios from 'axios';

export default function NewNonConformity({ history }) {

    const [departments, setDepartments] = useState([]);

    const [description, setDescription] = useState('');
    const [ocurrenceDate, setOcurrenceDate] = useState('');
    const [selectedDepartments, setSelectedDepartments] = useState([]);

    useEffect(() => {
        Axios.get('http://localhost:3000/departments')
            .then(res => {
                setDepartments(res.data);
            });
    }, []);

    const handleSubmit = e => {
        e.preventDefault();

        let checkedDepartments = [];

        departments.map(item => {
            if (item.isChecked)
                checkedDepartments.push(item.id);
        });

        const object = {
            description,
            ocurrenceDate,
            departments: checkedDepartments
        }

        Axios.post('http://localhost:3000/non-conformities', object)
            .then(res => {
                console.log(res)
            });
    }

    const handleDepartmentCheckbox = (e, id) => {
        let temp = [...departments];
        temp.find((item) => {
            return item.id === id;
        }).isChecked = e.target.checked;
        setDepartments(temp);
    }

    return (
        <>
            <div>
                <button onClick={() => { history.goBack() }}>Voltar</button>
            </div>
            <div>
                <form id='new-nonconformity-form' onSubmit={handleSubmit}>
                    <div>
                        <label>
                            Descrição
                        <input type="text" id="description" onChange={e => setDescription(e.target.value)} />
                        </label>
                    </div>
                    <div>
                        <label>
                            Data da Ocorrência
                            <input type="date" id="ocurrence-date" onChange={e => setOcurrenceDate(e.target.value)} />
                        </label>
                    </div>
                    <div>
                        {departments.map(item =>
                            <label key={item.id}>
                                {item.name}
                                <input type="checkbox" checked={item.isChecked} onChange={e => handleDepartmentCheckbox(e, item.id)} />
                            </label>
                        )}
                    </div>
                    <button type="submit">
                        Enviar
                    </button>
                </form>
            </div>
        </>
    );
}