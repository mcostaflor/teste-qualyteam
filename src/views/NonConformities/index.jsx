import React, { useEffect, useState } from 'react';

import {
    Card,
    CardContent,
    CardActions,
    Button,
    Typography,
    Chip,
    IconButton,
    TextField
} from '@material-ui/core';

import FilterListIcon from '@material-ui/icons/FilterList';
import AddIcon from '@material-ui/icons/Add';

import {
    nonConformitiesApi
} from '../../services/api/routes';

export default function NonConfirmities({ history }) {

    const [nonConformities, setNonConformities] = useState([]);
    const [sortAsc, setSortAsc] = useState(false);
    const [searchString, setSearchString] = useState('');

    useEffect(() => {
        async function fetchNonConformities() {
            setNonConformities(await nonConformitiesApi.getAll());
        }
        fetchNonConformities();
    }, []);

    const toggleSort = function () {
        setSortAsc(!sortAsc);
    }

    const filterNonConformitiesByString = function (list) {
        return list.filter(item => {
            return item.description.toUpperCase().includes(searchString.toUpperCase());
        });
    }

    const sortNonConformitiesByDate = function (list) {
        return list.sort((a, b) => {
            if (a && b) {
                let partsA = a['ocurrence-date'].split('-');
                let partsB = b['ocurrence-date'].split('-');
                if (new Date(`${partsA[1]}-${partsA[0]}-${partsA[2]}`) > new Date(`${partsB[1]}-${partsB[0]}-${partsB[2]}`))
                    return (sortAsc ? 1 : -1)
                else
                    return (sortAsc ? -1 : 1)
            }
            return 0;
        });
    }

    return (
        <>
            <div>
                <IconButton onClick={() => history.push('/nonconformities/new')} variant='contained' color={'primary'}>
                    <AddIcon />
                </IconButton>
                <IconButton onClick={toggleSort}>
                    <FilterListIcon style={{ transform: sortAsc ? 'rotate(180deg)' : 'rotate(0deg)' }} />
                </IconButton>
            </div>
            <TextField
                fullWidth
                variant={"outlined"}
                label={"Buscar..."}
                value={searchString}
                onChange={e => setSearchString(e.target.value)}
            />
            {filterNonConformitiesByString(sortNonConformitiesByDate(nonConformities))
                .map((item) =>
                    <Card key={item.id} style={{
                        marginTop: 16
                    }}>
                        <CardContent>
                            <Typography variant="h6" style={{ display: 'flex', justifyContent: 'space-between', marginBottom: 8 }}>
                                <div>
                                    #{item.id}
                                </div>
                                <div>
                                    {item['ocurrence-date']}
                                </div>
                            </Typography>
                            <Typography variant="body2" component="p">
                                {item.description}
                            </Typography>
                        </CardContent>
                        <CardContent>
                            {item._departments.length > 0 && item._departments.map(item =>
                                <Chip key={item.id} label={item.name} style={{ marginRight: 8 }} />
                            )}
                        </CardContent>
                        <CardActions>
                            <Button onClick={() => history.push(`nonconformities/${item.id}`)} variant={'outlined'} color="primary">
                                Ver
                        </Button>
                        </CardActions>
                    </Card>
                )}
        </>
    );
}