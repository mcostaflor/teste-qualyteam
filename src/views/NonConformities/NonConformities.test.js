import { render, cleanup, waitForElement } from '@testing-library/react'

import NonConformities from './index';
import renderApp from '../../helpers/tests/renderApp';

afterEach(cleanup);

describe('Testing non conformities screen', () => {

    it("renders", async () => {
        const page = render(renderApp(NonConformities, "/nonconformities", "/nonconformities", {}));
        expect(page).toMatchSnapshot();
    });

    it("loads a nonConformity", async () => {
        const { getByText } = render(renderApp(NonConformities, "/nonconformities", "/nonconformities", {}));

        const id = await waitForElement(() => getByText('#1'));

        expect(id).toBeDefined();

    })

});