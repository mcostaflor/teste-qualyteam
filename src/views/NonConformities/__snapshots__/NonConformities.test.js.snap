// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing non conformities screen renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
          tabindex="0"
          type="button"
          variant="contained"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              style="transform: rotate(0deg);"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          Buscar...
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-80 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-82"
            >
              <span>
                Buscar...
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
        tabindex="0"
        type="button"
        variant="contained"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            style="transform: rotate(0deg);"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Buscar...
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-80 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-82"
          >
            <span>
              Buscar...
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
